### PR TITLE
Turn i += 1 into ++i

### PR DESCRIPTION
**Pull request by fabiosantoscode**
This time I've added a check that ensures that the variable being incremented is at least numberish.

### CODE REVIEW

File: lib/compress.js [hunk 1]
```diff
@@ -4174,8 +4174,25 @@ merge(Compressor.prototype, {
         }
     }
 
+    function use_increment(self) {
+        if (self.step instanceof AST_Assign
+            && self.step.operator === "+="
+            && self.step.left instanceof AST_Symbol
+            && self.step.right.print_to_string() === "1"
```
Comment (kzc) [inline lib/compress.js]: `print_to_string` is very inefficient. Please check the specific AST node type - AST_Constant with value === 1.

File: lib/compress.js [hunk 2]
```diff
@@ -4174,8 +4174,25 @@ merge(Compressor.prototype, {
...
+            && self.init instanceof AST_Definitions
+            && self.init.definitions[0].name.name === self.step.left.name
+            && !isNaN(self.init.definitions[0].value.getValue())
```
Comment (kzc) [inline lib/compress.js]: That's pretty obscure/cryptic - please use this instead:
```
             && typeof self.init.definitions[0].value.getValue() == "number"
```

File: lib/compress.js [hunk 3]
```diff
@@ -4174,8 +4174,25 @@ merge(Compressor.prototype, {
         }
     }
 
+    function use_increment(self) {
+        if (self.step instanceof AST_Assign
+            && self.step.operator === "+="
+            && self.step.left instanceof AST_Symbol
+            && self.step.right.print_to_string() === "1"
+            && self.init instanceof AST_Definitions
+            && self.init.definitions[0].name.name === self.step.left.name
```
[the remark this answers is not in the document]
Comment (kzc) [reply lib/compress.js]: For example, if the loop variable is reassigned in the loop body to a non-number this code will produce the wrong result.

**Comment by kzc** [comment]
Expected:
```
$ echo 'for (var i = 0; i < 5; i += 1){ i += "" } console.log(i);' | node
011
```
This PR:
```
$ echo 'for (var i = 0; i < 5; i += 1){ i += "" } console.log(i);' | terser -c | node
5
```
```
$ echo 'for (var i = 0; i < 5; i += 1){ i += "" } console.log(i);' | terser -c
for(var i=0;i<5;++i)i+="";console.log(i);
```
Correctly optimizing loop increment in the general case is more difficult than you think.

To be honest, all this complexity to save one byte with the risk of potentially introducing wrong results is not worth it.

**Comment by kzc** [comment]
To further complicate the solution, a given `var` can also be legally redeclared multiple times with a function - even within the same for init.

Expected:
```
$ echo 'for (var i = 0, s = "A", i = "0"; i < 5; i += 1){ s += i; } console.log(i, s);' | node
011 A001
```
This PR:
```
$ echo 'for (var i = 0, s = "A", i = "0"; i < 5; i += 1){ s += i; } console.log(i, s);' | terser -c | node
5 'A01234'
```
It's not worth it.

**Comment by fabiosantoscode** [comment]
But nobody would normally redeclare a for loop variable, especially one that's incremented specifically with a number -- it's safe!

**Comment by kzc** [comment]
> But nobody would normally redeclare a for loop variable, especially one that's incremented specifically with a number -- it's safe!

That's not the standard of quality we expect from Uglify and now with Terser. The minifier should be able to handle any type of code - however implausible - and correctly produce the result.

If shortcuts are taken that affect reliability I would not be interested in contributing to this project any longer.

**Comment by fabiosantoscode** [comment]
Let's ensure terser is strict, then.

**Pull request closed without merging**